sdk/src/main/scala/io/horizen/account/state/ForgerStakeV2MsgProcessor.scala

```diff
@@ -0,0 +1,588 @@
+    requireIsNotPayable(invocation)
+    checkForgerStakesV2IsActive(view)
+
+    val inputParams = getArgumentsFromData(invocation.input)
+    val WithdrawCmdInput(forgerPublicKeys, stakedAmount) = WithdrawCmdInputDecoder.decode(inputParams)
+
+    log.debug(s"withdraw called - $forgerPublicKeys $stakedAmount")
+
+    if (stakedAmount.signum() != 1) {
+      val msg = s"Withdrawal amount must be greater than zero: $stakedAmount"
+      log.debug(msg)
+      throw new ExecutionRevertedException(msg)
+    }
+
+    if (!isValidZenAmount(stakedAmount)) {
+      val msg = s"Value is not a legal wei amount: $stakedAmount"
+      log.debug(msg)
+      throw new ExecutionRevertedException(msg)
+    }
+
+    val epochNumber = context.blockContext.consensusEpochNumber
+
+    StakeStorage.removeStake(view, forgerPublicKeys.blockSignPublicKey, forgerPublicKeys.vrfPublicKey,
+      epochNumber, invocation.caller, stakedAmount)
+
+    val withdrawStakeEvt = WithdrawForgerStake(invocation.caller, forgerPublicKeys.blockSignPublicKey, forgerPublicKeys.vrfPublicKey, stakedAmount)
+    val evmLog = getEthereumConsensusDataLog(withdrawStakeEvt)
+    view.addLog(evmLog)
+
+    view.subBalance(contractAddress, stakedAmount)
+    view.addBalance(invocation.caller, stakedAmount)
+
+    Array.emptyByteArray
+  }
+
+  def checkForgerStakesV2IsActive(view: BaseAccountStateView): Unit = {
+    if (!StakeStorage.isActive(view)) {
+      val msg = "Forger stake V2 has not been activated yet"
+      log.debug(msg)
+      throw new ExecutionRevertedException("Forger stake V2 has not been activated yet")
+    }
+  }
+
+  def doStakeTotalCmd(invocation: Invocation, view: BaseAccountStateView, currentEpoch: Int): Array[Byte] = {
+    requireIsNotPayable(invocation)
+    checkForgerStakesV2IsActive(view)
+
+    val inputParams = getArgumentsFromData(invocation.input)
+    val cmdInput = StakeTotalCmdInputDecoder.decode(inputParams)
+
+    val forgerKeys = cmdInput.forgerPublicKeys
+    val delegator = cmdInput.delegator
+    val consensusEpochStart = if (cmdInput.consensusEpochStart.isEmpty) currentEpoch else cmdInput.consensusEpochStart.get
+    val maxNumOfEpoch = cmdInput.maxNumOfEpoch
+    log.info(s"stakeTotal called - $forgerKeys $delegator epochStart: $consensusEpochStart - maxNumOfEpoch: $maxNumOfEpoch")
+
+    if (forgerKeys.isEmpty && delegator.isDefined) {
+      val msgStr = s"Illegal argument - delegator is defined while forger keys are not"
+      throw new ExecutionRevertedException(msgStr)
+    }
+    val consensusEpochEnd =
+      if (maxNumOfEpoch.isEmpty) consensusEpochStart
+      else if (consensusEpochStart + maxNumOfEpoch.get > currentEpoch) currentEpoch
+      else consensusEpochStart + maxNumOfEpoch.get - 1
+
+    val response: StakeTotalCmdOutput = StakeStorage.getStakeTotal(view, forgerKeys, delegator, consensusEpochStart, consensusEpochEnd)
```

Comment:
In case the forger doesn't exist or the delegator haven't delegated to this forger, this method returns the same a list of stake of maxNumOfEpoch length. Maybe it should perform a check before and throw an exception as in doRewardsReceivedCmd.